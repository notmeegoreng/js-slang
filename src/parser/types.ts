import type { Program } from 'estree'

import type { Context, Node, Chapter, Variant, SourceError } from '../types'

export type { ParserOptions as BabelOptions } from '@babel/parser'
export type { Options as AcornOptions } from 'acorn'

export interface Parser<TOptions> {
  parse(
    programStr: string,
    context: Context,
    options?: Partial<TOptions>,
    throwOnError?: boolean
  ): Program | null
  validate(ast: Program, context: Context, throwOnError?: boolean): boolean
}
export interface Rule<T extends Node> {
  /**
   * Name of the rule
   */
  name: string

  /**
   * Test snippets to test the behaviour of the rule. Providing no test snippets
   * means that this rule will not be tested when running unit tests.\
   * First element of the tuple is the code to test. Set the second element to `undefined`
   * if the snippet should not throw an error. Otherwise set it to the `explain()` value
   * of the error.
   */
  testSnippets?: [code: string, expected: string | undefined][]

  /**
   * Disable this rule for this chapter (inclusive) and above
   */
  disableFromChapter?: Chapter
  disableForVariants?: Variant[]
  checkers: {
    [name: string]: (node: T, ancestors: Node[]) => SourceError[]
  }
}
