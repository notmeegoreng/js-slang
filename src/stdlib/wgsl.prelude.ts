export const wgslPrelude = `
function silence_sound(duration) {
  return make_sound((_t) => 0, duration);
}

function noise_sound(duration){
  return make_sound((t) => _random() * 2 - 1, duration);
}

function consecutively(list_of_sounds) {
  function consec_two(ss1, ss2) {
    const wave1 = get_wave(ss1);
    const wave2 = get_wave(ss2);
    const dur1 = get_duration(ss1);
    const dur2 = get_duration(ss2);
    const new_wave = (t) => (t < dur1 ? wave1(t) : wave2(t - dur1));
    return make_sound(new_wave, dur1 + dur2);
  }
  return accumulate(consec_two, silence_sound(0), list_of_sounds);
}

function simultaneously(list_of_sounds) {
  function simul_two(ss1, ss2) {
    const wave1 = get_wave(ss1);
    const wave2 = get_wave(ss2);
    const dur1 = get_duration(ss1);
    const dur2 = get_duration(ss2);
    // new_wave assumes sound discipline (ie, wave(t) = 0 after t > dur)
    const new_wave = (t) => wave1(t) + wave2(t);
    // new_dur is higher of the two dur
    const new_dur = dur1 < dur2 ? dur2 : dur1;
    return make_sound(new_wave, new_dur);
  }

  const mushed_sounds = accumulate(simul_two, silence_sound(0), list_of_sounds);
  const normalised_wave = (t) => get_wave(mushed_sounds)(t) / length(list_of_sounds);
  const highest_duration = get_duration(mushed_sounds);
  return make_sound(normalised_wave, highest_duration);
}

function sine_sound(freq, duration) {
  return make_sound((t) => math_sin(2 * math_PI * t * freq), duration);
}

function square_sound(f, duration) {
  const fourier_expansion_square = t => accumulate(
    (x, y) => y + math_sin(2 * math_PI * (2 * x - 1) * f * t) / (2 * x - 1), 
    0, 
    list(1, 2, 3, 4, 5));
  return make_sound((t) => (4 / math_PI) * fourier_expansion_square(t), duration);
}

function linear_decay(decay_period) {
  return (t) => 
  t > decay_period || t < 0
  ? 0
  : 1 - t / decay_period;
}

function adsr(
  attack_ratio,
  decay_ratio,
  sustain_level,
  release_ratio
) {
  return (sound) => {
    const wave = get_wave(sound);
    const duration = get_duration(sound);
    const attack_time = duration * attack_ratio;
    const decay_time = duration * decay_ratio;
    const release_time = duration * release_ratio;
    
    return make_sound((x) => 
      x < attack_time
      ? wave(x) * (x / attack_time)
      : x < attack_time + decay_time
      ? ((1 - sustain_level) * linear_decay(decay_time)(x - attack_time)
            + sustain_level)
          * wave(x)
      : x < duration - release_time
      ? wave(x) * sustain_level
      : wave(x)
          * sustain_level
          * linear_decay(release_time)(x - (duration - release_time))
    , duration);
  };
}

function letter_name_to_frequency(note) {
  return midi_note_to_frequency(letter_name_to_midi_note(note));
}

function letter_name_to_midi_note(note) {
  return note === 'A0' ? 21 :
         note === 'A0#' || note === 'B0b' ? 22 :
         note === 'B0' ? 23 :

         note === 'C1' ? 24 :
         note === 'C1#' || note === 'D1b' ? 25 :
         note === 'D1' ? 26 :
         note === 'D1#' || note === 'E1b' ? 27 :
         note === 'E1' ? 28 :
         note === 'F1' ? 29 :
         note === 'F1#' || note === 'G1b' ? 30 :
         note === 'G1' ? 31 :
         note === 'G1#' || note === 'A1b' ? 32 :
         note === 'A1' ? 33 :
         note === 'A1#' || note === 'B1b' ? 34 :
         note === 'B1' ? 35 :

         note === 'C2' ? 36 :
         note === 'C2#' || note === 'D2b' ? 37 :
         note === 'D2' ? 38 :
         note === 'D2#' || note === 'E2b' ? 39 :
         note === 'E2' ? 40 :
         note === 'F2' ? 41 :
         note === 'F2#' || note === 'G2b' ? 42 :
         note === 'G2' ? 43 :
         note === 'G2#' || note === 'A2b' ? 44 :
         note === 'A2' ? 45 :
         note === 'A2#' || note === 'B2b' ? 46 :
         note === 'B2' ? 47 :

         note === 'C3' ? 48 :
         note === 'C3#' || note === 'D3b' ? 49 :
         note === 'D3' ? 50 :
         note === 'D3#' || note === 'E3b' ? 51 :
         note === 'E3' ? 52 :
         note === 'F3' ? 53 :
         note === 'F3#' || note === 'G3b' ? 54 :
         note === 'G3' ? 55 :
         note === 'G3#' || note === 'A3b' ? 56 :
         note === 'A3' ? 57 :
         note === 'A3#' || note === 'B3b' ? 58 :
         note === 'B3' ? 59 :

         note === 'C4' ? 60 :
         note === 'C4#' || note === 'D4b' ? 61 :
         note === 'D4' ? 62 :
         note === 'D4#' || note === 'E4b' ? 63 :
         note === 'E4' ? 64 :
         note === 'F4' ? 65 :
         note === 'F4#' || note === 'G4b' ? 66 :
         note === 'G4' ? 67 :
         note === 'G4#' || note === 'A4b' ? 68 :
         note === 'A4' ? 69 :
         note === 'A4#' || note === 'B4b' ? 70 :
         note === 'B4' ? 71 :

         note === 'C5' ? 72 :
         note === 'C5#' || note === 'D5b' ? 73 :
         note === 'D5' ? 74 :
         note === 'D5#' || note === 'E5b' ? 75 :
         note === 'E5' ? 76 :
         note === 'F5' ? 77 :
         note === 'F5#' || note === 'G5b' ? 78 :
         note === 'G5' ? 79 :
         note === 'G5#' || note === 'A5b' ? 80 :
         note === 'A5' ? 81 :
         note === 'A5#' || note === 'B5b' ? 82 :
         note === 'B5' ? 83 :

         note === 'C6' ? 84 :
         note === 'C6#' || note === 'D6b' ? 85 :
         note === 'D6' ? 86 :
         note === 'D6#' || note === 'E6b' ? 87 :
         note === 'E6' ? 88 :
         note === 'F6' ? 89 :
         note === 'F6#' || note === 'G6b' ? 90 :
         note === 'G6' ? 91 :
         note === 'G6#' || note === 'A6b' ? 92 :
         note === 'A6' ? 93 :
         note === 'A6#' || note === 'B6b' ? 94 :
         note === 'B6' ? 95 :

         note === 'C7' ? 96 :
         note === 'C7#' || note === 'D7b' ? 97 :
         note === 'D7' ? 98 :
         note === 'D7#' || note === 'E7b' ? 99 :
         note === 'E7' ? 100 :
         note === 'F7' ? 101 :
         note === 'F7#' || note === 'G7b' ? 102 :
         note === 'G7' ? 103 :
         note === 'G7#' || note === 'A7b' ? 104 :
         note === 'A7' ? 105 :
         note === 'A7#' || note === 'B7b' ? 106 :
         note === 'B7' ? 107 :

         note === 'C8' ? 108 : 0;
}

function stacking_adsr(
  waveform,
  base_frequency,
  duration,
  envelopes
) {
  const zip = (lst, n) => 
    is_null(lst)
    ? lst
    : pair(pair(n, head(lst)), zip(tail(lst), n + 1));

  return simultaneously(
    accumulate(
      (x, y) => pair(tail(x)(waveform(base_frequency * head(x), duration)), y),
      null,
      zip(envelopes, 1)
    )
  );
}

function midi_note_to_frequency(note) {
  return 440 * math_pow(2, (note - 69) / 12);
}

function cello(note, duration) {
  return stacking_adsr(
    square_sound,
    midi_note_to_frequency(note),
    duration,
    list(adsr(0.05, 0, 1, 0.1), adsr(0.05, 0, 1, 0.15), adsr(0, 0, 0.2, 0.15))
  );
}

function bell(note, duration) {
  return stacking_adsr(
    square_sound,
    midi_note_to_frequency(note),
    duration,
    list(
      adsr(0, 0.6, 0, 0.05),
      adsr(0, 0.6618, 0, 0.05),
      adsr(0, 0.7618, 0, 0.05),
      adsr(0, 0.9071, 0, 0.05)
    )
  );
}

function phase_mod(freq, duration, amount) {
  return (modulator) => make_sound(
    (t) => math_sin(2 * math_PI * t * freq + amount * get_wave(modulator)(t)),
    duration
  );
}

function triangle_sound(freq, duration) {
  const fourier_expansion_triangle = t => accumulate(
    (x, y) => y + (math_pow(-1, x) * math_sin((2 * x + 1) * t * freq * math_PI * 2))
        / math_pow(2 * x + 1, 2), 
    0, 
    list(1, 2, 3, 4, 5));
  return make_sound(
      (t) => (8 / math_PI / math_PI) * fourier_expansion_triangle(t),
      duration);
}

function piano(note, duration) {
  return stacking_adsr(
    triangle_sound,
    midi_note_to_frequency(note),
    duration,
    list(adsr(0, 0.515, 0, 0.05), adsr(0, 0.32, 0, 0.05), adsr(0, 0.2, 0, 0.05))
  );
}

function play_wave(wave, duration) {
  return play(make_sound(wave, duration));
}

function sawtooth_sound(freq, duration) {
  const fourier_expansion_sawtooth = t => accumulate(
    (x, y) => y +  math_sin(2 * math_PI * x * freq * t) / x, 
    0, 
    list(1, 2, 3, 4, 5));
  return make_sound(
      (t) => 1 / 2 - (1 / math_PI) * fourier_expansion_sawtooth(t),
      duration);
}

function violin(note, duration) {
  return stacking_adsr(
    sawtooth_sound,
    midi_note_to_frequency(note),
    duration,
    list(
      adsr(0.35, 0, 1, 0.15),
      adsr(0.35, 0, 1, 0.15),
      adsr(0.45, 0, 1, 0.15),
      adsr(0.45, 0, 1, 0.15)
    )
  );
}

function trombone(note, duration) {
  return stacking_adsr(
    square_sound,
    midi_note_to_frequency(note),
    duration,
    list(adsr(0.2, 0, 1, 0.1), adsr(0.3236, 0.6, 0, 0.1))
  );
}
`
