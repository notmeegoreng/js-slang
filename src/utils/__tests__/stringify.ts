import { list, set_tail, tail } from '../../stdlib/list'
import { Chapter, type Value } from '../../types'
import { stripIndent } from '../formatters'
import { lineTreeToString, stringDagToLineTree, stringify, valueToStringDag } from '../stringify'
import { expectResult, testMultipleCases } from '../testing'

type TestCase = [desc: string, valueToStringify: Value, expected: string]
const cases: TestCase[] = [
  // Primitives
  ['String representation of numbers are nice', 0, '0'],
  ['String representation of string are nice', 'a string', '"a string"'],
  ['String representation of booleans are nice', true, 'true'],
  [
    'String representation of arrow functions are nice',
    // @ts-ignore
    (x, y) => x,
    '(x, y) => x'
  ],
  ['String representation of null is nice', null, 'null'],
  ['String representation of undefined is nice', undefined, 'undefined'],

  // Arrays
  [
    'String representations of arrays are nice',
    [1, 'true', true, () => 1],
    '[1, "true", true, () => 1]'
  ],
  [
    'String representation of multidimensional arrays are nice',
    [1, 'true', [true, () => 1, [[]]]],
    '[1, "true", [true, () => 1, [[]]]]'
  ],
  ['String representation of empty arrays are nice', [], '[]'],
  ['String representation works with arrays with holes 1', [, []], '[undefined, []]'],
  ['String representation works with arrays with holes 2', [, , []], '[undefined, undefined, []]'],
  [
    'String representation of huge arrays is nice',
    Array.from(Array(100)).map((_, i) => i),
    stripIndent`
            [ 0,
              1,
              2,
              3,
              4,
              5,
              6,
              7,
              8,
              9,
              10,
              11,
              12,
              13,
              14,
              15,
              16,
              17,
              18,
              19,
              20,
              21,
              22,
              23,
              24,
              25,
              26,
              27,
              28,
              29,
              30,
              31,
              32,
              33,
              34,
              35,
              36,
              37,
              38,
              39,
              40,
              41,
              42,
              43,
              44,
              45,
              46,
              47,
              48,
              49,
              50,
              51,
              52,
              53,
              54,
              55,
              56,
              57,
              58,
              59,
              60,
              61,
              62,
              63,
              64,
              65,
              66,
              67,
              68,
              69,
              70,
              71,
              72,
              73,
              74,
              75,
              76,
              77,
              78,
              79,
              80,
              81,
              82,
              83,
              84,
              85,
              86,
              87,
              88,
              89,
              90,
              91,
              92,
              93,
              94,
              95,
              96,
              97,
              98,
              99]
      `
  ],

  // Lists
  [
    'String representtion of lists are nice',
    list(1, 2, 3, 4, 5, 6, 7, 8, 9, 10),
    '[1, [2, [3, [4, [5, [6, [7, [8, [9, [10, null]]]]]]]]]]'
  ],
  [
    'String representation of huge lists are nice',
    list(...Array.from(Array(100)).map((_, i) => i + 1)),
    stripIndent`
            [ 1,
            [ 2,
            [ 3,
            [ 4,
            [ 5,
            [ 6,
            [ 7,
            [ 8,
            [ 9,
            [ 10,
            [ 11,
            [ 12,
            [ 13,
            [ 14,
            [ 15,
            [ 16,
            [ 17,
            [ 18,
            [ 19,
            [ 20,
            [ 21,
            [ 22,
            [ 23,
            [ 24,
            [ 25,
            [ 26,
            [ 27,
            [ 28,
            [ 29,
            [ 30,
            [ 31,
            [ 32,
            [ 33,
            [ 34,
            [ 35,
            [ 36,
            [ 37,
            [ 38,
            [ 39,
            [ 40,
            [ 41,
            [ 42,
            [ 43,
            [ 44,
            [ 45,
            [ 46,
            [ 47,
            [ 48,
            [ 49,
            [ 50,
            [ 51,
            [ 52,
            [ 53,
            [ 54,
            [ 55,
            [ 56,
            [ 57,
            [ 58,
            [ 59,
            [ 60,
            [ 61,
            [ 62,
            [ 63,
            [ 64,
            [ 65,
            [ 66,
            [ 67,
            [ 68,
            [ 69,
            [ 70,
            [ 71,
            [ 72,
            [ 73,
            [ 74,
            [ 75,
            [ 76,
            [ 77,
            [ 78,
            [ 79,
            [ 80,
            [ 81,
            [ 82,
            [ 83,
            [ 84,
            [ 85,
            [ 86,
            [ 87,
            [ 88,
            [89, [90, [91, [92, [93, [94, [95, [96, [97, [98, [99, [100, null]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]]
           `
  ],

  // Objects
  [
    'String representation of objects is nice',
    { a: 1, b: true, c: () => 1 },
    '{"a": 1, "b": true, "c": () => 1}'
  ],
  [
    'String representation of nested objects is nice',
    { a: 1, b: true, c: () => 1, d: { e: 5, f: 6 } },
    '{"a": 1, "b": true, "c": () => 1, "d": {"e": 5, "f": 6}}'
  ],
  [
    'String representation of big objects are nice',
    {
      a: 1,
      b: true,
      c: () => 1,
      d: { e: 5, f: 6 },
      g: 0,
      h: 0,
      i: 0,
      j: 0,
      k: 0,
      l: 0,
      m: 0,
      n: 0
    },
    stripIndent`
            { "a": 1,
              "b": true,
              "c": () => 1,
              "d": {"e": 5, "f": 6},
              "g": 0,
              "h": 0,
              "i": 0,
              "j": 0,
              "k": 0,
              "l": 0,
              "m": 0,
              "n": 0}
          `
  ]
]

testMultipleCases(cases, ([value, expected]) => {
  expect(stringify(value)).toEqual(expected)
})

// Test cases that are a little bit more complicated
test('Correctly handles circular lists with multiple entry points', () => {
  const x = list(1, 2, 3)
  set_tail(tail(tail(x)), x)
  const value = stringify(list(x, tail(x), tail(tail(x))))
  expect(value).toEqual(
    stripIndent`[ [1, [2, [3, ...<circular>]]],\n[[2, [3, [1, ...<circular>]]], [[3, [1, [2, ...<circular>]]], null]]]`
  )
})

test('String representation of functions are nice', () => {
  // @ts-ignore
  function f(x, y) {
    return x
  }

  return expect(stringify(f)).toMatchInlineSnapshot(`
      "function f(x, y) {
        return x;
      }"
    `)
})

test('String representation of non literal objects is nice', () => {
  const errorMsg: string = 'This is an error'
  const errorObj: Error = new Error(errorMsg)
  return expect(stringify(errorObj)).toMatchInlineSnapshot(`"${errorObj.toString()}"`)
})

test('String representation of non literal objects in nested object is nice', () => {
  const errorMsg: string = 'This is an error'
  const errorObj: Error = new Error(errorMsg)
  const nestedObj: Object = {
    data: [1, [2, errorObj], 3]
  }
  return expect(stringify(nestedObj)).toMatchInlineSnapshot(
    `"{\\"data\\": [1, [2, ${errorObj.toString()}], 3]}"`
  )
})

test('String representation of objects with circular references is nice', () => {
  let o: any = {}
  o.o = o
  expect(stringify(o)).toEqual('{"o": ...<circular>}')
})

test('String representation of objects with toReplString member calls toReplString', () => {
  const toReplString = jest.fn(() => '<RUNE>')
  const o = { toReplString }
  expect(stringify(o)).toEqual('<RUNE>')
  expect(toReplString).toHaveBeenCalledTimes(1)
})

test('String representation of builtins are nice', () => {
  return expectResult(
    stripIndent`
  stringify(pair);
  `,
    Chapter.SOURCE_2
  ).toMatchInlineSnapshot(`
            "function pair(left, right) {
            	[implementation hidden]
            }"
          `)
})

test('String representation with 1 space indent', () => {
  return expectResult(
    stripIndent`
  stringify(parse('x=>x;'), 1);
  `,
    Chapter.SOURCE_4
  ).toMatchInlineSnapshot(`
            "[\\"lambda_expression\\",
            [[[\\"name\\", [\\"x\\", null]], null],
            [[\\"return_statement\\", [[\\"name\\", [\\"x\\", null]], null]], null]]]"
          `)
})

test('String representation with default (2 space) indent', () => {
  return expectResult(
    stripIndent`
  stringify(parse('x=>x;'));
  `,
    Chapter.SOURCE_4
  ).toMatchInlineSnapshot(`
            "[ \\"lambda_expression\\",
            [ [[\\"name\\", [\\"x\\", null]], null],
            [[\\"return_statement\\", [[\\"name\\", [\\"x\\", null]], null]], null]]]"
          `)
})

test('String representation with more than 10 space indent should trim to 10 space indent', () => {
  return expectResult(
    stripIndent`
  stringify(parse('x=>x;'), 100);
  `,
    Chapter.SOURCE_4
  ).toMatchInlineSnapshot(`
            "[         \\"lambda_expression\\",
            [         [[\\"name\\", [\\"x\\", null]], null],
            [[\\"return_statement\\", [[\\"name\\", [\\"x\\", null]], null]], null]]]"
          `)
})

test('lineTreeToString', () => {
  return expect(
    lineTreeToString({
      type: 'block',
      prefixFirst: '[ ',
      prefixRest: '  ',
      block: [
        {
          type: 'block',
          prefixFirst: '[ ',
          prefixRest: '  ',
          block: [
            { type: 'line', line: { type: 'terminal', str: 'why', length: 3 } },
            { type: 'line', line: { type: 'terminal', str: 'hello', length: 5 } }
          ],
          suffixRest: ',',
          suffixLast: ' ]'
        },
        { type: 'line', line: { type: 'terminal', str: 'there', length: 5 } },
        { type: 'line', line: { type: 'terminal', str: 'sethbling here', length: 42 } }
      ],
      suffixRest: ',',
      suffixLast: ' ]'
    })
  ).toMatchInlineSnapshot(`
            "[ [ why,
                hello ],
              there,
              sethbling here ]"
          `)
})

test('stringDagToLineTree', () => {
  return expect(
    lineTreeToString(
      stringDagToLineTree(
        {
          type: 'multiline',
          lines: ['hello world', 'why hello there', "it's a", '  multiline', 'string!'],
          length: 42
        },
        2,
        80
      )
    )
  ).toMatchInlineSnapshot(`
            "hello world
            why hello there
            it's a
              multiline
            string!"
          `)
})

test('stringDagToLineTree part 2', () => {
  return expect(
    stringDagToLineTree(
      {
        type: 'pair',
        head: { type: 'terminal', str: '42', length: 2 },
        tail: {
          type: 'pair',
          head: { type: 'terminal', str: '69', length: 2 },
          tail: { type: 'terminal', str: 'null', length: 4 },
          length: 42
        },
        length: 42
      },
      2,
      80
    )
  ).toMatchInlineSnapshot(`
            Object {
              "line": Object {
                "head": Object {
                  "length": 2,
                  "str": "42",
                  "type": "terminal",
                },
                "length": 42,
                "tail": Object {
                  "head": Object {
                    "length": 2,
                    "str": "69",
                    "type": "terminal",
                  },
                  "length": 42,
                  "tail": Object {
                    "length": 4,
                    "str": "null",
                    "type": "terminal",
                  },
                  "type": "pair",
                },
                "type": "pair",
              },
              "type": "line",
            }
          `)
})

test('stringDagToLineTree part 3', () => {
  return expect(
    lineTreeToString(
      stringDagToLineTree(
        {
          type: 'pair',
          head: { type: 'terminal', str: '42', length: 2 },
          tail: {
            type: 'pair',
            head: { type: 'terminal', str: '69', length: 2 },
            tail: { type: 'terminal', str: 'null', length: 4 },
            length: 42
          },
          length: 42
        },
        2,
        80
      )
    )
  ).toMatchInlineSnapshot(`"[42, [69, null]]"`)
})

test('stringDagToLineTree part 4', () => {
  return expect(
    lineTreeToString(
      stringDagToLineTree(
        {
          type: 'pair',
          head: { type: 'terminal', str: '42', length: 2 },
          tail: {
            type: 'pair',
            head: { type: 'terminal', str: '69', length: 2 },
            tail: { type: 'terminal', str: 'null', length: 4 },
            length: 42
          },
          length: 99
        },
        2,
        80
      )
    )
  ).toMatchInlineSnapshot(`
            "[ 42,
            [69, null]]"
          `)
})

test('value to StringDag', () => {
  return expect(
    lineTreeToString(
      stringDagToLineTree(
        valueToStringDag([
          1,
          [
            2,
            [
              3,
              [
                4,
                [
                  5,
                  [
                    6,
                    [
                      7,
                      [
                        8,
                        [9, [10, [11, [12, [13, [14, [15, [16, [17, [18, [19, [20, null]]]]]]]]]]]]
                      ]
                    ]
                  ]
                ]
              ]
            ]
          ]
        ]),
        2,
        80
      )
    )
  ).toMatchInlineSnapshot(`
            "[ 1,
            [ 2,
            [ 3,
            [ 4,
            [ 5,
            [ 6,
            [ 7,
            [8, [9, [10, [11, [12, [13, [14, [15, [16, [17, [18, [19, [20, null]]]]]]]]]]]]]]]]]]]]"
          `)
})
