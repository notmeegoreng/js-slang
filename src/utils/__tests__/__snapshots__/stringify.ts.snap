// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`String representation of builtins are nice: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "stringify(pair);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "function pair(left, right) {
	[implementation hidden]
}",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`String representation with 1 space indent: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "stringify(parse('x=>x;'), 1);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "[\\"lambda_expression\\",
[[[\\"name\\", [\\"x\\", null]], null],
[[\\"return_statement\\", [[\\"name\\", [\\"x\\", null]], null]], null]]]",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`String representation with default (2 space) indent: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "stringify(parse(\\"x=>x;\\"));",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "[ \\"lambda_expression\\",
[ [[\\"name\\", [\\"x\\", null]], null],
[[\\"return_statement\\", [[\\"name\\", [\\"x\\", null]], null]], null]]]",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`String representation with more than 10 space indent should trim to 10 space indent: expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "stringify(parse(\\"x=>x;\\"), 100);",
  "displayResult": Array [],
  "numErrors": 0,
  "parsedErrors": "",
  "result": "[         \\"lambda_expression\\",
[         [[\\"name\\", [\\"x\\", null]], null],
[[\\"return_statement\\", [[\\"name\\", [\\"x\\", null]], null]], null]]]",
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;
